datafusion/core/src/execution/context.rs

```diff
@@ -1528,6 +1528,9 @@ impl SessionState {
             // repartitioning and local sorting steps to meet distribution and ordering requirements.
             // Therefore, it should run before EnforceDistribution and EnforceSorting.
             Arc::new(JoinSelection::new()),
+            // Enforce sort before PipelineFixer
```

Comment:
Why doess the enforcement happens before pipeline fixer (and not for example, also before `JoinSelection`)?

Reply:
We haven't focused on optimal rule ordering yet. Right now, we only enable SHJ when joining two unbounded streams, so it has to happen at some point before `PipelineFixer`. As we mature the SHJ implementation, we will enable it even for normal tables when it is appropriate (yields performance gains). Within that context, we will revisit rule ordering.

Reply:
If the filter expression is incapable of pruning both sides, bounded sources will not experience any performance boost since there is no sliding window effect. We plan to include a check for whether the filter expression supports range pruning on both sides in a future PR. Once this check is in place, we can evaluate the feasibility of applying this plan to bounded sources.

Reply:
I think the comment for `PipelineFixer` need to be adjusted.

```rust
// If the query is processing infinite inputs, the PipelineFixer rule applies the
// necessary transformations to make the query runnable (if it is not already runnable).
// If the query can not be made runnable, the rule emits an error with a diagnostic message.
// Since the transformations it applies may alter output partitioning properties of operators
// (e.g. by swapping hash join sides), this rule runs before EnforceDistribution.
```